A high-level `ChatInterface` abstraction

abidlabs opened the pull request:
Simple examples of usage:

```py
import gradio as gr

def echo(message, history):
   return message

gr.ChatInterface(echo).launch()
```

```py
from gradio import ChatInterface
import time

def double(message, history):
    time.sleep(1)
    return message + message

ChatInterface(double, examples=["bbbb", "abcdef"], cache_examples=True).queue().launch()
```

```py
from gradio import ChatInterface
import time

def echo_stream(message, history):
    for i in range(len(message)):
        time.sleep(0.3)
        yield message[:i]

ChatInterface(echo_stream).queue().launch()
```

The UI:

<img width="1316" alt="image" src="https://github.com/gradio-app/gradio/assets/1778297/18cf337f-215c-43f9-87aa-4116d55af332">


We'll address in future PRs if there's demand:
* Handle async fns and async generators
* Convert submit button to stop button while streaming
* Add other helpful API routes


Closes: #3510

osanseviero reviewed (commented):
Cool stuff!

abidlabs commented on gradio/chat_interface.py:
```diff
@@ -0,0 +1,193 @@
+            chatbot: an instance of the gr.Chatbot component to use for the chat interface. If not provided, a default gr.Chatbot component will be created.
+            textbox: an instance of the gr.Textbox component to use for the chat interface. If not provided, a default gr.Textbox component will be created.
+            examples: sample inputs for the function; if provided, appear below the chatbot and can be clicked to populate the chatbot input.
+            cache_examples: If True, caches examples in the server for fast runtime in examples. The default option in HuggingFace Spaces is True. The default option elsewhere is False.
```
Thanks @osanseviero I'll take a look

abidlabs commented on gradio/chat_interface.py:
```diff
@@ -0,0 +1,268 @@
+    def _api_stream_fn(self, message: str, history: list[list[str]]):
+        history = self._display_input(message, history)
+        for response in self._stream_fn(message, history):
+            return response
```
```suggestion
            yield response
```

freddyaboulton commented on gradio/chat_interface.py:
```diff
@@ -0,0 +1,268 @@
+        # a chat history, instead it is just stored internally in the state.
+        history = State([])
+
+        self.fake_api_btn.click(
```
Something's not right here - the view api page and client both show that this route takes and returns nothing

<img width="405" alt="image" src="https://github.com/gradio-app/gradio/assets/41651716/9e25ee10-5978-4b48-a864-d7c0537506b4">

<img width="943" alt="image" src="https://github.com/gradio-app/gradio/assets/41651716/9d4e733d-35de-4c7b-84cb-1effaa92dd2c">

This is with this demo

```python
from gradio import ChatInterface
import time

def double(message, history):
    time.sleep(1)
    return message + message

ChatInterface(double, examples=["bbbb", "abcdef"], cache_examples=True).queue().launch()
```

abidlabs replied:
Should be fixed now, thanks @freddyaboulton!

freddyaboulton commented:
Thanks @abidlabs ! Will take a closer look tomorrow

freddyaboulton commented on gradio/chat_interface.py:
```diff
@@ -0,0 +1,268 @@
+        self.examples = examples
+        self.cache_examples = cache_examples
+        self.buttons: list[Button] = []
+        self.history = []
```
nit: this isn't used anywhere

abidlabs replied:
Removed

freddyaboulton commented on gradio/chat_interface.py:
```diff
@@ -0,0 +1,269 @@
+    def _api_stream_fn(self, message: str, history: list[list[str]]):
+        history = self._display_input(message, history)
+        for response in self.fn(message, history):
+            history[-1][1] = response
```
history can actually be a `List[Tuple[str]]` which will error out when trying to do the item assignment. This is caused by the client since it converts everything to tuples (I think we should remove in 4.0). 


Repro


<img width="657" alt="image" src="https://github.com/gradio-app/gradio/assets/41651716/b3c67570-22fa-493d-ab0c-830dbc4c6cfa">

```bash
Traceback (most recent call last):
  File "/home/user/.local/lib/python3.10/site-packages/gradio/routes.py", line 439, in run_predict
    output = await app.get_blocks().process_api(
  File "/home/user/.local/lib/python3.10/site-packages/gradio/blocks.py", line 1384, in process_api
    result = await self.call_function(
  File "/home/user/.local/lib/python3.10/site-packages/gradio/blocks.py", line 1103, in call_function
    prediction = await utils.async_iteration(iterator)
  File "/home/user/.local/lib/python3.10/site-packages/gradio/utils.py", line 343, in async_iteration
    return await iterator.__anext__()
  File "/home/user/.local/lib/python3.10/site-packages/gradio/utils.py", line 336, in __anext__
    return await anyio.to_thread.run_sync(
  File "/home/user/.local/lib/python3.10/site-packages/anyio/to_thread.py", line 33, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
  File "/home/user/.local/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 877, in run_sync_in_worker_thread
    return await future
  File "/home/user/.local/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 807, in run
    result = context.run(func, *args)
  File "/home/user/.local/lib/python3.10/site-packages/gradio/utils.py", line 319, in run_sync_iterator_async
    return next(iterator)
  File "/home/user/.local/lib/python3.10/site-packages/gradio/utils.py", line 688, in gen_wrapper
    yield from f(*args, **kwargs)
  File "/home/user/.local/lib/python3.10/site-packages/gradio/chat_interface.py", line 258, in _api_stream_fn
    history[-1][1] = response
TypeError: 'tuple' object does not support item assignment
```

abidlabs replied:
Nice catch

freddyaboulton commented on gradio/chat_interface.py:
```diff
@@ -0,0 +1,269 @@
+                    self.buttons.append(submit_btn)
+                self.stop_btn = Button("Stop", variant="stop", visible=False)
+                
+                for btn in [retry_btn, delete_last_btn, clear_btn]:
```
nit: why not use a gr.ClearButton here?

abidlabs replied:
Sure, will update

abidlabs replied:
Hmm it actually complicates things a bit because we have to separately handle the case if users supply a button (in which case we still have to add the event handlers manually). I'll just leave as is

freddyaboulton commented on gradio/chat_interface.py:
```diff
@@ -0,0 +1,269 @@
+                    self.textbox = Textbox(show_label=False, placeholder="Type a message...")
+            
+            with Row():
+                if submit_btn:
```
I think the UI would look a lot sleeker if the submit button was next to the textbox

<img width="944" alt="Screenshot 2023-07-13 at 10 57 11 AM" src="https://github.com/gradio-app/gradio/assets/41651716/d666d81b-3c3e-486a-8f0c-885372448cd1">

Similar to ChatGPT, HuggingChat, and iOS

abidlabs replied:
Ok let me play around with that and see how it looks on different screen sizes

freddyaboulton commented on gradio/chat_interface.py:
```diff
@@ -0,0 +1,269 @@
+        history.append((message, None))
+        return history
+
+    def _submit_fn(self, message: str, history: list[list[str]]):
```
So the history contains the user message (since we call _display_history)? I think that's confusing. I would expect the history to only contain the history up to but not including the current message.

abidlabs replied:
Yeah its confusing since In `_display_input`, `history` contains the history not including the current message, but by the time that `_submit_fn` is called, `history` includes the current message. These are all internal methods so I don't think it matters too much, but I'll rename `history` to be `history_with_input` or something

freddyaboulton replied:
I'm worried this will introduce bugs into demos built with ChatInterface. I think it would be natural for a developer to do something like the following (pseudocode):

```python
def chat(message, history):
    chain.history = history
    return chain.predict(message)
```
And then pass duplicate entries for `message` to the chain. BTW I think there is an inconsistency between `_api_stream_fn` and `_api_fn` the latter does not include the latest message in the chain but the former does.

Whatever we pick we should be consistent but I think it's more natural for the history to not include the latest message.

abidlabs replied:
Oh wait I made a mistake in `_submit_fn`. The history should absolutely NOT include the last message. I misunderstood what you were saying earlier. Sorry, I'll fix and test this out

abidlabs commented:
Thanks for the initial review. I'm going to be working on addressing these comments & finishing off the remaining items

abidlabs commented:
Feedback from @yvrjsharma along with my comments:

> I wondered why we need to explicitly mention the delete button name to include it in the UI. While it's not the case with Retry and clear buttons.

Yeah I think I'll make the delete button appear by default, especially if we bring the submit button to the same line as the Textbox

> I think the docstring for ChatInterface says that we can include a Chatbot component as a parameter, or one would be included by default. I was trying to understand in which cases would we need to explicitly mention the chatbot param?

If you want to customize the Chatbot in any way (e.g. the height or the latex_delimiters). I'll explain this in the docstring

> I also found it difficult to understand how to pass parameters like temperature, max_tokens, top_p, etc., while calling my inference (predict) function.

Yeah this isn't currently possible, and I agree its a pretty big limitation. I'm thinking we can have an additional `inputs` parameter that takes inputs beyond the textbox and passes them into your functions. My main question is where should these questions show up in the UI?

> One strange behavior I noticed was that on Colab, the initial value of chatbot comes in as a blank list [], while with the exact same code on Spaces, the initial value captured is chatbot = [[first_user_input, None]]  I had to handle this when I moved my code from Colabs to a Space.

That seems weird, can you share a minimal code example?

> Till now, we used to add the bot response to the chatbot list (a list of sub-lists of user and corresponding bot messages) and use to return that to the output component (chatbot). However, now we just need to return the bot_response, and the interface takes care of everything else. I really liked this part as it reduced the number of lines of code on my end, reducing the chances of error.

🔥

abidlabs commented:
@freddyaboulton the issues you pointed out should be fixed. Here's example code:

```py
from gradio import ChatInterface
import random

def test(message, history):
    return f"{len(history)+1} messages so far. Here's a random number: {random.random()}"

ChatInterface(test).launch()
```

Tested with the Client as well:

```py
from gradio_client import Client

client = Client(url)

client.reset_session()
client.predict("test")
```


Testing streaming right now

abidlabs commented:
Streaming is working too. Test code:

```py
from gradio import ChatInterface
import random
import time

def test(message, history):
    message = f"{len(history)+1} messages so far. Here's a random number: {random.random()}"
    for i in range(len(message)):
        time.sleep(0.01)
        yield message[:i]

_, url, _ = ChatInterface(test).queue().launch()


# from gradio_client import Client

# client = Client(url)

# for x in client.submit("test"):
#     print(x)
```

abidlabs commented:
@yvrjsharma your first two points have been addressed as well.

Working on tests now

freddyaboulton commented:
> My main question is where should these questions show up in the UI?

I like the parameters accordion in falcon chat demo: https://huggingface.co/spaces/HuggingFaceH4/falcon-chat

Something we can think about for the future - how to make our high-level abstractions more modular? If you want to use ChatInterface for convenience but your use case isn't 100% covered you have to use gr.Blocks() from scratch.  Maybe we can let the function be None and let users configure it themselves if they know what they're doing. This example introduces an `add_fn` method (not advocating for that api just pointing out what I think would be cool to do)

```python
chat = gr.ChatInterface(fn=None, delete_last_btn="del").queue()

def predict_with_params(msg, history, temperature):
    # do cool stuff
    pass

def handle_upload(file, msg, history):
    # As a question about the uploaded file
    pass

with gr.Blocks() as demo:
    chat.render()
    with gr.Row():
        with gr.Column():
            with gr.Accordion(label="Parameters"):
                temp = gr.Slider()
        with gr.Column():
            upload = gr.UploadButton()
    chat.add_fn(predict_with_params, inputs=[temp])
    chat.add_fn(handle_upload, inputs=[upload])
```

Something to think about for later.

yvrjsharma commented:
> My main question is where should these questions show up in the UI?
I agree with @freddyaboulton's suggestion on this. An accordion below the buttons would make a good resting place.

abidlabs commented:
@yvrjsharma can't reproduce this locally (with and without queuing), testing on Spaces rn

yvrjsharma commented:
Thanks for checking this, @abidlabs. I now understand this was because of the different wheels between us.

abidlabs commented:
After merging in the the changes from the `fix_group` branch, the UI looks like this:

<img width="1174" alt="image" src="https://github.com/gradio-app/gradio/assets/1778297/72f7b9be-096f-4582-a328-42191d2a206d">

freddyaboulton reviewed (commented):
This is looking really good! Noticed a typo in the guide lol

abidlabs commented:
Thanks @freddyaboulton! Still a fair bit of to do work to do on the docs and the functional test, but feel free to review everything else

abidlabs commented:
Ok all right, tests are in place. I'm going to make one more pass through the docs and then this should be ready

abidlabs commented:
> I also found it difficult to understand how to pass parameters like temperature, max_tokens, top_p, etc., while calling my inference (predict) function.

I have a simple idea to address this. We could add an optional `additional_inputs` parameter to `gr.ChatInterface()` which accepts a list of components. The values of these components would be passed into your chat function after `message` and `history`. You could create the components anywhere within a larger `gr.Blocks()` context

Example usage:

```py
def chat(message, history, system_prompt):
   ...

with gr.Blocks() as demo:
     prompt = gr.Textbox()
     gr.ChatInterface(chat, additional_inputs=prompt)
demo.launch()
```

You could still do the accordion, but you'd just create it yourself (which I think is better than forcing this particular UI design)

taoari commented:
@abidlabs This is nice work, it makes Chatbot much easier to use!

I would suggest adding an upload button on the left of the input box, and with some callback like:

```py
def _upload_file(msg, filepath):
    # history = history + [((file.name,), None)]
    mtype = mimetypes.guess_type(filepath.name)[0]
    if mtype.startswith('image'):
        msg += f'<img src="\\file={filepath.name}" alt="{os.path.basename(filepath.name)}"/>'
    elif mtype.startswith('audio'):
        msg += f'<audio controls><source src="\\file={filepath.name}">{os.path.basename(filepath.name)}</audio>'
    elif mtype.startswith('video'):
        msg += f'<video controls><source src="\\file={filepath.name}">{os.path.basename(filepath.name)}</video>'
    else:
        msg += f'<a href="\\file={filepath.name}">📁 {os.path.basename(filepath.name)}</a>'
    return msg

upload = gr.UploadButton("📁", file_types=["image", "video", "audio", "file"])
upload.upload(_upload_file, [msg, upload], [msg], queue=False)
```

for basic multimodal input support.

taoari commented:
@abidlabs This is nice work, it makes Chatbot much easier to use!

I would suggest adding an upload button on the left of the input box, and with some callback like:

```py
def _upload_file(msg, filepath):
    # history = history + [((file.name,), None)]
    mtype = mimetypes.guess_type(filepath.name)[0]
    if mtype.startswith('image'):
        msg += f'<img src="\\file={filepath.name}" alt="{os.path.basename(filepath.name)}"/>'
    elif mtype.startswith('audio'):
        msg += f'<audio controls><source src="\\file={filepath.name}">{os.path.basename(filepath.name)}</audio>'
    elif mtype.startswith('video'):
        msg += f'<video controls><source src="\\file={filepath.name}">{os.path.basename(filepath.name)}</video>'
    else:
        msg += f'<a href="\\file={filepath.name}">📁 {os.path.basename(filepath.name)}</a>'
    return msg

upload = gr.UploadButton("📁", file_types=["image", "video", "audio", "file"])
upload.upload(_upload_file, [msg, upload], [msg], queue=False)
```

for basic multimodal input support.

taoari commented:
Another thing is that I found the current implementation is not streaming friendly. For example, you have to use the following code to stream the chat message:

```py
  def echo(message, history):
      for i, c in enumerate(message):
          yield message[:i+1]
```

It would be better if the following callback function would work:

```py
def echo(message, history):
    for c in message:
        yield c
```

Finally, I would suggest "Delete Last" could be changed to "Undo". When undo, the last message should also come back to the textbox. In this case, "Retry" may be removed, as it is simply an "Undo"+"Submit". "Clear History" may also be renamed to "Clear" for simplicity. The UI may be like the following to save space:

![image](https://github.com/gradio-app/gradio/assets/3815006/45c08fd0-0323-4da9-be44-7be3cbd7e9ac)

freddyaboulton commented on guides/04_chatbots/01_creating-a-chatbot-fast.md:
````diff
@@ -0,0 +1,254 @@
+
+Of course, why use a closed-source model when you can use an open-source one instead? Here's the equivalent example using Together's RedePajama model.
+
+```py
````
Should we deploy as a space so that we can embed it in the docs and have users try it out? WDYT @yvrjsharma ?

freddyaboulton commented on gradio/chat_interface.py:
```diff
@@ -0,0 +1,326 @@
+                if chatbot:
+                    self.chatbot = chatbot.render()
+                else:
+                    self.chatbot = Chatbot(label="Input")
```
Shouldn't the default label for the chatbot be "chat" or "chatbot" as opposed to "Input"?

abidlabs commented:
Thanks for the suggestions @taoari! Here are my thoughts:

> I would suggest adding an upload button on the left of the input box, and with some callback like:

We'll add multimodal support in a future PR (probably something like what you proposed)

> Finally, I would suggest "Delete Last" could be changed to "Undo". When undo, the last message should also come back to the textbox. In this case, "Retry" may be removed, as it is simply an "Undo"+"Submit". "Clear History" may also be renamed to "Clear" for simplicity. The UI may be like the following to save space and fit into single row:

I like "Undo"! Will update the UI. Not sure if putting all of the buttons in the same line will look good, but will play around with this

> Another thing is that I found the current implementation is not streaming friendly. For example, you have to use the following code to stream the chat message:

There are two reasons for this approach:

1. Its consistent with how generator functions work in `gr.Interface()` and `gr.Blocks()`
2. Its more general than just sending the individual chunks back because e.g. it allows the chatbot streaming function to "edit" previous responses. 

I do see the point in having a simpler interface like you've suggested, perhaps that could be toggled via a parameter at some point, but not the default.

aliabd commented:
@abidlabs i made some fixes to the website stuff, please take a look at the preview and lmk what you think.

abidlabs commented:
Thanks @aliabd perfect!

abidlabs commented:
I've addressed the feedback from @freddyaboulton and switched from "delete previous" to "undo" and updated the tests accordingly. Here's how the UI currently looks:

<img width="1156" alt="image" src="https://github.com/gradio-app/gradio/assets/1778297/47539051-75b9-4154-95e3-0a0e39f43a76">


Will discuss internally on whether to move the buttons to a single line or not.

taoari commented:
In streaming mode, one has to add `yield ""` in front of the bot function, otherwise, an error will occur if the user presses ENTER with empty input. Anyway to fix this?

```py
from gradio import ChatInterface
import time

def echo_stream(message, history):
    yield ""
    for i in range(len(message)):
        time.sleep(0.3)
        yield message[:i+1]

ChatInterface(echo_stream).queue().launch()
```

abidlabs commented:
Thanks @taoari for catching that. I've handled this edge case so that if a generator does not `yield` anything, it is automatically assumed to `yield None` and the chatbot does not error out (in the UI, no bot response apperas). That's analogous to a regular function not having a return statement which is assumed to `return None` in Python.
